Add metrics_rs integration for task metrics

jkosh44 opened the pull request:
This commit provides integration of task metrics and metrics.rs.

Resolves #71

I hope that no one was already working on this, I saw the issue was unassigned so I thought I'd give it a try.

I split the PR into two commits for easier reviewing. The first commit moves shared macros and traits to a separate module, but doesn't change any code (other than imports, exports, and visibility modifiers). The second commit actually adds the task metrics integration.

There is still some duplicate across the `XMetricsReporterBuilder` and `XMetricsReporter` structs. I played around with making a generic struct that can be used by both, but the structs themselves are fairly simple, so it didn't seem worth the added complexity. There is a decent amount of duplicated doc-comments across the structs though. Let me know if you have any thoughts.

jkosh44 commented on src/metrics_rs.rs:
```diff
@@ -0,0 +1,221 @@
+            }
+        }
+    }
+}
```
To be honest, I was a little confused by this `impl`. It seems like it's specific to a single histogram metric in the runtime. Does this make it impossible to add any other histogram metrics in the future?

arielb1 commented on src/metrics_rs.rs:
```diff
@@ -0,0 +1,221 @@
+
+macro_rules! metric_refs {
+    (
+        [$struct_name:ident] [$($ignore:ident),* $(,)?] [$metrics_name:ty] [$emit_arg_type:ty] {
```
yea, better to pass the identifiers explicitly rather than constructing them

arielb1 commented on src/task/metrics_rs_integration.rs:
```diff
@@ -0,0 +1,290 @@
+/// The published metrics are the fields of [`TaskMetrics`], but with the
+/// `tokio_` prefix added, for example, `instrumented_count`. If you have multiple [`TaskMonitor`]s
+/// then it is strongly recommended to use [`with_metrics_transformer`] to give each [`TaskMonitor`]
+/// a unique metric name.
```
This definitely feels to me like a bad-ish API for the gauges, since if you have multiple `TaskMonitor`s they will probably conflict in an annoying way.

jkosh44 replied:
I've removed the default constructor in my most recent commit and make the metrics transformer mandatory.

jkosh44 commented:
@arielb1 thanks so much for the review! I believe I've addressed all your comments in my most recent commit. Please let me know if I've missed something.

arielb1 commented on src/task/metrics_rs_integration.rs:
```diff
@@ -0,0 +1,293 @@
+    }
+}
+
+impl Default for TaskMetricsReporterBuilder {
```
You should remove this?
